Implement ELLA for SD1.5

ily-R opened the pull request:
This is my draft PR for the bounty **ELLA for SD1.5**. I implemented and tested the architecture. I have few steps left.

- [x]  Implement ELLA
- [x] Test the adapter's inject/eject 
- [x] implement the weight conversion script for these [weights](https://huggingface.co/QQGYLab/ELLA/discussions/5)
- [x] test ELLA contribution to SD1.5 model using **t5_xl** text_embeddings from [here](https://huggingface.co/google/flan-t5-xl)

limiteinductive commented on src/refiners/foundationals/latent_diffusion/ella_adapter.py:
```diff
@@ -0,0 +1,294 @@
+    pass
+
+
+class _PerceiverAttentionBlock(fl.Chain):
+    def __init__(
+        self,
+        width: int,
+        num_heads: int,
+        timestep_embedding_dim: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            fl.Sum(
+                fl.Parallel(
+                    fl.GetArg(0),
+                    fl.Chain(
+                        fl.Distribute(
+                            LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                            InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                        ),
+                        fl.Parallel(
+                            fl.GetArg(0),
+                            fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
+                            fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
+                        ),
+                        fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype),
+                    ),
+                )
+            ),
+            fl.Residual(
+                fl.Chain(
+                    OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                    fl.Linear(width, width * 4, device=device, dtype=dtype),
+                    SquaredReLU(),
+                    fl.Linear(width * 4, width, device=device, dtype=dtype),
+                )
+            ),
+        )
```
There are quite a few things that need to be corrected here. fl.Sum(fl.Parallel(*args)) is the same as fl.Sum(*args). This is because when the input of a sum is a tuple, it will sum it, but you can also pass multiple layers. Also  fl.Chain(fl.Sum(*args1), fl.Residual(*args2)) is the same thing that fl.Sum(*(args1 + args2)). So you can simplify everything and use a fl.Sum. It should look like this (haven't tested it):
```
class _PerceiverAttentionBlock(fl.Sum):
    def __init__(
        self,
        width: int,
        num_heads: int,
        timestep_embedding_dim: int,
        device: Device | str | None = None,
        dtype: DType | None = None,
    ) -> None:
        super().__init__(
            fl.GetArg(0),
            fl.Chain(
                fl.Distribute(
                    LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                    InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                ),
                fl.Parallel(
                    fl.GetArg(0),
                    fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
                    fl.Concatenate(fl.GetArg(0), fl.GetArg(1), dim=1),
                ),
                fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype),
            ),
            fl.Chain(
                OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.Linear(width, width * 4, device=device, dtype=dtype),
                SquaredReLU(),
                fl.Linear(width * 4, width, device=device, dtype=dtype),
            ),
        )
  ```
  
Also, you should refrain from using GetArg too much. I think you should put the variable GetArg(1) in the context and remove all GetArg(0).

ily-R replied:
You are right passing one of the inputs as a context will make the code readable and avoid using GetArg(). I ended up with the following:
`
class PerceiverAttentionBlock(fl.Chain):
    def __init__(
        self,
        width: int,
        num_heads: int,
        timestep_embedding_dim: int,
        device: Device | str | None = None,
        dtype: DType | None = None,
    ) -> None:
        super().__init__(
            fl.Passthrough(
                fl.UseContext("ella_perceiverresampler", "text_input"),
                InputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.SetContext("ella_perceiverresampler", "proj_text_input"),
            ),
            fl.Residual(
                LatentsAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                fl.Parallel(
                    fl.Identity(),
                    fl.Concatenate(fl.Identity(), fl.UseContext("ella_perceiverresampler", "proj_text_input"), dim=1),
                    fl.Concatenate(fl.Identity(), fl.UseContext("ella_perceiverresampler", "proj_text_input"), dim=1),
                ),
                fl.Attention(embedding_dim=width, num_heads=num_heads, device=device, dtype=dtype)
            ),
            fl.Residual(
                    OutputAdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
                    fl.Linear(width, width * 4, device=device, dtype=dtype),
                    SquaredReLU(),
                    fl.Linear(width * 4, width, device=device, dtype=dtype),
            ),
        )
`
what do you think ?

ily-R replied:
the context "text_input" is set at the beginning of PerceiverResampler

ily-R replied:
following Cedric comments, ill refactor the implementation using usecontext() and following the perceiverresampler implementation on ip-adapter.

limiteinductive reviewed (changes requested):
Thanks for the great work.

I made my first round of essentially stylistic nits. I haven't made an e2e test, but verified that conversion is working.

Could you also apply lint using `rye fmt` and verify typing with `rye run pyright`?

deltheil commented on src/refiners/foundationals/latent_diffusion/ella_adapter.py:
```diff
@@ -0,0 +1,294 @@
+        )
+
+
+class PerceiverResampler(fl.Chain):
```
Out of curiosity: did you have look at the Perceiver Resampler already implemented part of IP-Adapter Plus? see https://github.com/finegrain-ai/refiners/blob/10fba3467ad3cd9e878ad3b7e797b2804cbbb261/src/refiners/foundationals/latent_diffusion/image_prompt.py#L182

Are there noticeable differences? If yes, are there ways to keep this new version consistent with the other one?

ily-R replied:
I did check this implementation. except for `PerceiverScaledDotProductAttention`  They are identical actually. I tried to follow the same modulation from the official ELLA repo. But I can reformulate my perceiverResampler implementation to match the one for IP-adapter.

deltheil replied:
> They are identical actually with few differences

OK, good to know.

In other similar cases (you might have noticed that there are several ViT-s version in Refiners - for CLIP image encoder, DINOv2, etc) we decided to have multiple versions (no DRY) precisely because there are some subtleties here and there.

> But I can reformulate my perceiverResampler implementation to match the one for IP-adapter.

I think that would be great (at least, it's worth trying!). Thanks!

ily-R replied:
got it, ill commit the new implementation

ily-R replied:
```
            Latents(num_latents, width, device=device, dtype=dtype),
            fl.Residual(
                fl.UseContext("ella", "timestep_embedding"),
                fl.SiLU(),
                fl.Linear(time_embedding_dim, width, device=device, dtype=dtype),
            ),
```
after checking, I found that the difference between ip perceiverresampler and ella perceiverresampler is the use of the time embeddings, they first sum the Latents with a feedforward projection of the time_embedding. They also use the time_embedding component with LayerNorm in the AdaLayerNorm module. And finally they use a squared Relu.

limiteinductive commented on scripts/conversion/convert_ella_adapter.py:
```diff
@@ -0,0 +1,92 @@
+        required=True,
+        dest="source_path",
+        help=("Official checkpoint from https://huggingface.co/QQGYLab/ELLA/blob/main/ella-sd1.5-tsc-t5xl.safetensors"),
+    )
```
You can use this: https://huggingface.co/docs/huggingface_hub/guides/download

ily-R replied:
done

Laurent2916 commented on src/refiners/foundationals/latent_diffusion/ella_adapter.py:
```diff
@@ -0,0 +1,285 @@
+        return x * (1 + scale) + shift
+
+
+class ParameterInitialized(fl.Parameter):
+    def __init__(
+        self, *dims: int, requires_grad: bool = True, device: Device | str | None = None, dtype: DType | None = None
+    ) -> None:
+        super().__init__(*dims, requires_grad=requires_grad, device=device, dtype=dtype)
+        nn.init.normal_(self.weight, mean=0, std=dims[1] ** 0.5)
+
+
+class Latents(fl.Chain):
+    def __init__(
+        self,
+        num_latents: int,
+        width: int,
+        device: Device | str | None = None,
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            ParameterInitialized(
+                num_latents,
+                width,
+                device=device,
+                dtype=dtype,
+            ),
+        )
```
nit: I'm wondering if `ParameterInitialized` and `Latents` could be fused together in a single class. Since `Latents` is simply a chain with a single element: `ParameterInitialized`.

ily-R replied:
I tried to follow the same thing you did for DINOV2 with ClassToken

Laurent2916 replied:
ah right, we do this because this looks prettier when printing the model's repr, see https://github.com/finegrain-ai/refiners/pull/132#discussion_r1423714231

Laurent2916 commented on src/refiners/foundationals/latent_diffusion/ella_adapter.py:
```diff
@@ -0,0 +1,285 @@
+        dtype: DType | None = None,
+    ) -> None:
+        super().__init__(
+            fl.Distribute(
+                AdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+                AdaLayerNorm(width, timestep_embedding_dim, device=device, dtype=dtype),
+            ),
+            fl.Parallel(
+                fl.GetArg(index=1),
+                fl.Lambda(func=self.to_kv),
+                fl.Lambda(func=self.to_kv),
+            ),
```
nit: slightly hard to read at a glance, could it be simplified ?

ily-R replied:
following Cedric comments, this is almost and identical implementation to IP adapter's PerceiverAttention class. Do you have a suggestion it simplify it ?

Laurent2916 replied:
I guess this version is the less verbose one then

Laurent2916 reviewed (commented):
Hey @ily-R, this looks better!
Do you see anything else @limiteinductive ?
I'll run the tests tomorrow

Laurent2916 commented:
I've run the tests, beside the fact that `sentencepiece` was missing from the dependencies, the 3 tests are green.

limiteinductive reviewed (commented):
I have nothing to add (except linting and rebase); @deltheil, could you have a final look? Also, what about the `sentencepiece` dependency?

deltheil commented:
> I have nothing to add (except linting and rebase);

That's great!

> @deltheil, could you have a final look?

OK, stay tuned.

> Also, what about the `sentencepiece` dependency?

@limiteinductive where is it used? ("[...] beside the fact that sentencepiece was missing from the dependencies, the 3 tests are green")

Laurent2916 commented:
Hi @ily-R, final nits before we merge:
- could you please add `sentencepiece` as a test dependency
- because of a recent PR, there a small conflict in `src/refiners/foundationals/latent_diffusion/stable_diffusion_1/__init__.py`, could please resolve it

ily-R commented:
I rebased and did the necessary lint/format and added the sentencepiece as a test dependency . The test are green.
Let me know if there is something to fix

deltheil reviewed (commented):
@ily-R sorry for the delay - I sent a few final nit-s, do you mind having a look? Also, please rebase and squash into a single commit. Thanks!

ily-R commented on tests/e2e/test_diffusion.py:
```diff
@@ -1799,6 +1814,45 @@ def test_diffusion_textual_inversion_random_init(
+    test_device: torch.device,
+):
+    sd15 = sd15_std_float16
+    t5_encoder = T5TextEmbedder(pretrained_path=test_weights_path / "QQGYLab/T5XLFP16", max_length=128).to(
```
done

deltheil commented on tests/e2e/test_diffusion.py:
```diff
@@ -506,6 +512,29 @@ def lda_ft_mse_weights(test_weights_path: Path) -> Path:
 
+@pytest.fixture(scope="module")
+def ella_weights(test_weights_path: Path) -> tuple[Path, Path]:
+    ella_adapter_weights = test_weights_path / "ELLA-Adapter/ella-sd1.5-tsc-t5xl.safetensors"
```
```suggestion
    ella_adapter_weights = test_weights_path / "ELLA-Adapter" / "ella-sd1.5-tsc-t5xl.safetensors"
```

deltheil replied:
nit (mostly for consistency)

deltheil commented on tests/e2e/test_diffusion.py:
```diff
@@ -590,10 +619,6 @@ def sd15_std_sde(
 def sd15_std_float16(
     text_encoder_weights: Path, lda_weights: Path, unet_weights_std: Path, test_device: torch.device
 ) -> StableDiffusion_1:
-    if test_device.type == "cpu":
-        warn("not running on CPU, skipping")
-        pytest.skip()
-
```
Any reason to remove these lines? This is probably a debug leftover. Could you please undo these changes?

ily-R replied:
oh! that was a typo

deltheil commented on tests/adapters/test_ella_adapter.py:
```diff
@@ -0,0 +1,41 @@
+
+
+@no_grad()
+def test_scale(test_device: torch.device):
```
```suggestion
def test_ella_cross_attention(test_device: torch.device):
```

deltheil replied:
Since there is no scale involved (it's rather testing/inspecting that the adapter properly modified the cross-attentions)

deltheil reviewed (commented):
Please take a look (last round I guess)

deltheil commented:
@ily-R thanks for the quick turnaround. Some last suggestions (see above).

deltheil commented on tests/e2e/test_diffusion.py:
```diff
@@ -506,6 +512,29 @@ def lda_ft_mse_weights(test_weights_path: Path) -> Path:
+    if not ella_adapter_weights.is_file():
+        warn(f"could not find weights at {ella_adapter_weights}, skipping")
+        pytest.skip(allow_module_level=True)
+    t5xl_weights = test_weights_path / "QQGYLab/T5XLFP16"
```
```suggestion
    t5xl_weights = test_weights_path / "QQGYLab" / "T5XLFP16"
```

limiteinductive reviewed (approved):
LGTM
